typescript/infra/src/agents/agent.ts

```diff
@@ -28,6 +28,119 @@ export abstract class AgentKey {
   }
 }
 
+export class ReadOnlyAgentKey extends AgentKey {
+  private _identifier: string;
+  private _address: string;
+
+  constructor(
+    public environment: string,
+    public readonly role: KEY_ROLE_ENUM,
+    identifier: string,
+    address: string,
+    public readonly chainName?: ChainName,
+    public readonly index?: number,
+  ) {
+    super(environment, role, chainName, index);
+
+    this._identifier = identifier;
+    this._address = address;
+  }
+
+  /**
+   * Parses the identifier, deriving the environment, role, chain (if any), and index (if any)
+   * and constructs a ReadOnlyAgentKey.
+   * @param identifier The "identifier" of the key. This can come in a few different
+   * flavors, e.g.:
+   * alias/abacus-testnet2-key-kathy (<-- not specific to any chain)
+   * alias/abacus-testnet2-key-optimismkovan-relayer (<-- chain specific)
+   * alias/abacus-testnet2-key-alfajores-validator-0 (<-- chain specific and has an index)
+   * abacus-dev-key-kathy (<-- same idea as above, but without the `alias/` prefix if it's not AWS-based)
+   * @param address The address of the key.
+   * @returns A ReadOnlyAgentKey for the provided identifier and address.
+   */
+  static fromSerializedAddress(
+    identifier: string,
+    address: string,
+  ): ReadOnlyAgentKey {
+    const regex =
+      /.*abacus-([a-zA-Z0-9]+)-key-([a-zA-Z0-9]+)-?([a-zA-Z0-9]+)?-?([0-9]+)?/g;
+    const matches = regex.exec(identifier);
+    if (!matches) {
+      throw Error('Invalid identifier');
+    }
+    const environment = matches[1];
+
+    const assertRole = (roleStr: string) => {
```

Comment:
Nit, any particular reason you're defined these lamdbas inline? Not a problem but makes this function a bit harder to read

Reply:
moved :)